### PR TITLE
feature: add tasks hw6 8/11/2019

### PR DESCRIPTION
**Pull request by Serheev**

### CODE REVIEW

File: src/main/java/com/serheev/Immutables/Immutable.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
+package com.serheev.Immutables;
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/Immutables/Immutable.java]: Immutables -> immutables

File: src/main/java/com/serheev/Immutables/Immutable.java [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+    private final List<T> myListString;
+    private final HashMap<K, V> myMapInt;
+    private final HashMap<K, V> myMapString;
+
+    public String getName() {
+        return name;
+    }
+
+    public List<T> getMyListInt() {
+        return new ArrayList(myListInt);
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/Immutables/Immutable.java]: are you tested this? it's doesn't work

File: src/main/java/com/serheev/annotations/SomeVersion.java [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+
+import java.lang.annotation.ElementType;
+import java.lang.annotation.Retention;
+import java.lang.annotation.RetentionPolicy;
+import java.lang.annotation.Target;
+
+@Target(ElementType.FIELD)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface SomeVersion {
+    float version();
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/annotations/SomeVersion.java]: double

File: src/main/java/com/serheev/serializable/Order.java [hunk 4]
```diff
@@ -0,0 +1,76 @@
...
+    Order(int id, double sum, String text, List<String> items) {
+        this.id = id;
+        this.date = LocalDate.now();
+        this.sum = sum;
+        this.text = text;
+        this.items = items;
+    }
+
+    public Integer getId() {
+        return id;
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/Order.java]: this.id

File: src/main/java/com/serheev/serializable/Order.java [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+    }
+
+    public LocalDate getDate() {
+        return date;
+    }
+
+    public Double getSum() {
+        return sum;
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/Order.java]: this.sum

File: src/main/java/com/serheev/serializable/Runner.java [hunk 6]
```diff
@@ -0,0 +1,34 @@
...
+import java.io.IOException;
+import java.util.Arrays;
+import java.util.HashMap;
+import java.util.List;
+
+import static com.serheev.serializable.SerializeUtils.*;
+
+public class Runner {
+    public static void main(String[] args) throws IOException, ClassNotFoundException {
+        List<String> goodsListOne = Arrays.asList("item1", "item2", "item3");
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/Runner.java]: move these all to test class

File: src/main/java/com/serheev/serializable/SerializeUtils.java [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+import java.io.FileOutputStream;
+import java.io.FileInputStream;
+import java.io.IOException;
+import java.io.ObjectInputStream;
+import java.io.ObjectOutputStream;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Set;
+
+public class SerializeUtils {
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/SerializeUtils.java]: SerializationUtils

Comment (Aleksandr04) [reply src/main/java/com/serheev/serializable/SerializeUtils.java]: 1/ final class
2/ add private constructor to this class

File: src/main/java/com/serheev/serializable/SerializeUtils.java [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+        }
+        return null;
+    }
+
+    // Arrays
+    public static void serializeArrayInt(Integer[] a) throws IOException {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream("arrayIntData.bin"))) {
+            oos.writeObject(a);
+        } catch (IOException e) {
+            e.printStackTrace();
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/SerializeUtils.java]: e.printStackTrace();
don't use this, need to use sout("error........")

File: src/test/java/com/serheev/Immutables/ImmutableTest.java [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+    @Test
+    public void shouldBeImmutableListString() {
+        imObject.getMyListString().add("Three");
+        Assert.assertEquals(list2, imObject.getMyListString());
+        Assert.assertEquals(list2.hashCode(), imObject.getMyListString().hashCode());
+    }
+
+    @Test
+    public void getMyMapInt() {
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/Immutables/ImmutableTest.java]: bad name of test cases

File: src/test/java/com/serheev/serializable/SerializeUtilsTest.java [hunk 10]
```diff
@@ -0,0 +1,92 @@
...
+        //Arrays.stream(origin).forEach(e -> System.out.println(e));
+    }
+
+    @Test
+    public void shouldBeConsistentForList() throws IOException, ClassNotFoundException {
+        List<?> origin = list;
+        List<?> deserialazed = deserializeList();
+        Assert.assertEquals(origin, deserialazed);
+        Assert.assertEquals(origin.hashCode(), deserialazed.hashCode());
+        //origin.forEach(e -> System.out.println(e));
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/serializable/SerializeUtilsTest.java]: remove all comments

File: src/test/java/com/serheev/singletones/SingletonAnnotationReflectionTest.java [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+
+import com.serheev.annotations.SomeVersion;
+import org.junit.Assert;
+import org.junit.Test;
+
+import java.lang.annotation.Annotation;
+import java.lang.reflect.Field;
+
+public class SingletonAnnotationReflectionTest {
+    float beforeRandom;
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/singletones/SingletonAnnotationReflectionTest.java]: private

File: src/test/java/com/serheev/singletones/SingletonAnnotationReflectionTest.java [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+
+    @Test
+    public void shouldMakeRandomFloatValue() {
+        Singleton mySingleton = Singleton.getInstance();
+        try {
+            Field field = mySingleton.getClass().getDeclaredField("version");
+            field.setAccessible(true);
+            beforeRandom = (float) field.get(mySingleton);
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            e.printStackTrace();
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/singletones/SingletonAnnotationReflectionTest.java]: sout("error...............")

File: src/test/java/com/serheev/singletones/SingletonAnnotationReflectionTest.java [hunk 13]
```diff
@@ -0,0 +1,50 @@
...
+                }
+            }
+
+        } catch (ClassNotFoundException | IllegalAccessException e) {
+            e.printStackTrace();
+        }
+
+        Assert.assertFalse(beforeRandom == afterRandom);
+
+        //System.out.println("before - " + beforeRandom);
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/singletones/SingletonAnnotationReflectionTest.java]: remove this

File: src/test/java/com/serheev/singletones/SingletonReflectionTest.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+        } catch (NoSuchFieldException | IllegalAccessException e) {
+            e.printStackTrace();
+        }
+    }
+
+    @Test
+    public void shouldGetNewInstance() {
+        Singleton newSingleton1 = Singleton.getInstance();
+        //System.out.println("s1 - " + newSingleton1.counter);
+        //System.out.println("s1 - " + newSingleton1.hashCode());
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/singletones/SingletonReflectionTest.java]: remove this

File: src/main/java/com/serheev/Immutables/Immutable.java (hunk 2 above)
Comment (Serheev) [reply src/main/java/com/serheev/Immutables/Immutable.java]: Yes, I did, but it works. Doesn't work only if "return myListInt" But with return new ArrayList(myListInt) it works. Why did you questioned? Something is wrong with this?

File: src/main/java/com/serheev/immutables/Immutable.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+    private final List<T> myListString;
+    private final HashMap<K, V> myMapInt;
+    private final HashMap<K, V> myMapString;
+
+    public String getName() {
+        return name;
+    }
+
+    public List<T> getMyListInt() {
+        return new ArrayList(myListInt);
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/immutables/Immutable.java]: need to return clones too

File: src/main/java/com/serheev/serializable/SerializationUtils.java [hunk 16]
```diff
@@ -0,0 +1,105 @@
...
+
+    public static void serializeOrder(Order[] o) throws IOException {
+        try (ObjectOutputStream oos = new ObjectOutputStream(new FileOutputStream("orders.bin"))) {
+            oos.writeObject(o);
+        } catch (IOException e) {
+            System.err.println("Serialization error: " + e);
+        }
+    }
+
+    public static Order[] deserializeOrder() throws IOException, ClassNotFoundException {
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/SerializationUtils.java]: if this method has catch block, so why to throws there exceptions?

File: src/main/java/com/serheev/singletones/Singleton.java [hunk 17]
```diff
@@ -0,0 +1,37 @@
+package com.serheev.singletones;
+
+// Double Checked Locking & volatile Singleton
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/singletones/Singleton.java]: remove this)

File: src/test/java/com/serheev/serializable/OrderSerializeTest.java [hunk 18]
```diff
@@ -0,0 +1,57 @@
...
+        List<String> list1 = Arrays.asList("item1", "item2", "item3");
+        List<String> list2 = Arrays.asList("item4", "item5", "item6");
+        List<String> list3 = Arrays.asList("item7", "item8", "item9");
+
+        Order order1 = new Order(1001, 100, "Pay cash", list1);
+        Order order2 = new Order(1002, 200, "Pay card", list2);
+        Order order3 = new Order(1003, 300, "VIP", list3);
+
+        orders = new Order[]{order1, order2, order3};
+        serializeOrder(orders);
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/serializable/OrderSerializeTest.java]: move this line to new test case

File: src/test/java/com/serheev/serializable/OrderSerializeTest.java [hunk 19]
```diff
@@ -0,0 +1,57 @@
...
+    @Test
+    public void shouldBeConsistent() throws IOException, ClassNotFoundException {
+        Order[] origin = orders;
+        Order[] deserialazed = deserializeOrder();
+
+        Assert.assertEquals(origin.getClass(), deserialazed.getClass());
+        Assert.assertEquals(Arrays.toString(orders), Arrays.toString(deserializeOrder()));
+        Assert.assertTrue(Arrays.equals(origin, deserialazed));
+
+        int count = 0;
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/serializable/OrderSerializeTest.java]: what is the main reason of this count?

File: src/test/java/com/serheev/serializable/SerializationUtilsTest.java [hunk 20]
```diff
@@ -0,0 +1,88 @@
...
+        set.add("Value2");
+        set.add("Value3");
+        set.add("Value2");
+        serializeSet(set);
+
+        hashmap = new HashMap<>();
+        hashmap.put(1, "Value1");
+        hashmap.put(2, "Value2");
+        hashmap.put(3, "Value3");
+        serializeHashMap(hashmap);
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/serializable/SerializationUtilsTest.java]: move to test too

File: src/main/java/com/serheev/singletones/Singleton.java [hunk 21]
```diff
@@ -0,0 +1,37 @@
...
+// from https://habr.com/ru/post/129494/
+
+import com.serheev.annotations.SomeVersion;
+
+public class Singleton implements Runnable {
+    public static int counter = 0;
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/singletones/Singleton.java]: make this private and create getCounter()

File: src/main/java/com/serheev/singletones/Singleton.java [hunk 22]
```diff
@@ -0,0 +1,38 @@
+package com.serheev.singletones;
+
+import com.serheev.annotations.SomeVersion;
+
+public class Singleton implements Runnable {
+    private static int counter = 0;
+    private String name = "defaultName";
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/singletones/Singleton.java]: move "defaultName" to the place where you create Singleton

File: src/main/java/com/serheev/singletones/Singleton.java [hunk 23]
```diff
@@ -0,0 +1,38 @@
...
+    private String surname = "defaultSurname";
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/singletones/Singleton.java]: "defaultSurname" move too

File: src/test/java/com/serheev/immutables/ImmutableTest.java [hunk 24]
```diff
@@ -0,0 +1,76 @@
...
+import org.junit.Test;
+
+import java.util.ArrayList;
+import java.util.HashMap;
+import java.util.List;
+
+public class ImmutableTest {
+    private Immutable imObject;
+    private String name;
+    private List<Integer> list1;
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/immutables/ImmutableTest.java]: move it to each cases

File: src/test/java/com/serheev/immutables/ImmutableTest.java [hunk 25]
```diff
@@ -0,0 +1,76 @@
...
+        imObject.getMyListString().add("Three");
+        Assert.assertEquals(list2, imObject.getMyListString());
+        Assert.assertEquals(list2.hashCode(), imObject.getMyListString().hashCode());
+    }
+
+    @Test
+    public void shouldBeImmutableMapInt() {
+        imObject.getMyMapInt().put(3, 3000);
+        Assert.assertEquals(map1, imObject.getMyMapInt());
+        Assert.assertEquals(map1.hashCode(), imObject.getMyMapInt().hashCode());
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/immutables/ImmutableTest.java]: add static import

File: src/test/java/com/serheev/serializable/SerializationUtilsTest.java [hunk 26]
```diff
@@ -0,0 +1,81 @@
...
+import static com.serheev.serializable.SerializationUtils.serializeSet;
+import static com.serheev.serializable.SerializationUtils.deserializeSet;
+import static com.serheev.serializable.SerializationUtils.serializeHashMap;
+import static com.serheev.serializable.SerializationUtils.deserializeHashMap;
+
+public class SerializationUtilsTest {
+    private Integer[] array;
+    private List<String> list;
+    private Set<String> set;
+    private HashMap<Integer, String> hashmap;
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/serializable/SerializationUtilsTest.java]: HashMap -> Map

File: src/test/java/com/serheev/serializable/OrderSerializeTest.java [hunk 27]
```diff
@@ -0,0 +1,57 @@
...
+
+        Order order1 = new Order(1001, 100, "Pay cash", list1);
+        Order order2 = new Order(1002, 200, "Pay card", list2);
+        Order order3 = new Order(1003, 300, "VIP", list3);
+
+        orders = new Order[]{order1, order2, order3};
+    }
+
+    @Test
+    public void shouldBeConsistent() throws IOException, ClassNotFoundException {
```
Comment (Aleksandr04) [inline src/test/java/com/serheev/serializable/OrderSerializeTest.java]: remove throw everywhere

File: src/main/java/com/serheev/serializable/SerializationUtils.java [hunk 28]
```diff
@@ -0,0 +1,105 @@
...
+        } catch (IOException e) {
+            System.err.println("Serialization error: " + e);
+        }
+    }
+
+    public static Order[] deserializeOrder() {
+        try (ObjectInputStream ois = new ObjectInputStream(new FileInputStream("orders.bin"))) {
+            return (Order[]) ois.readObject();
+        } catch (IOException | ClassNotFoundException e) {
+            System.err.println("Deserialization error: " + e);
```
Comment (Aleksandr04) [inline src/main/java/com/serheev/serializable/SerializationUtils.java]: throw new RuntimeException() instead of return null

**Pull request merged**